src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+            std::string msg = "Cyclic dependencies were found when attempting to resolve units in model '" + model->name() + "'. The dependency loop is:\n";
+            std::string spacer = "    ";
+            for (auto &h : history) {
+                msg += spacer + "units '" + std::get<0>(h) + "' imports '" + std::get<1>(h) + "' from '" + std::get<2>(h);
+                spacer = "',\n    ";
+            }
+            msg += "'.";
+            auto issue = Issue::create();
+            issue->setDescription(msg);
+            issue->setLevel(libcellml::Issue::Level::WARNING);
+            issue->setModel(model);
+            mImporter->addIssue(issue);
+            std::vector<std::tuple<std::string, std::string, std::string>>().swap(history);
+            return false;
+        }
+    }
+    return resolveComponentImports(model, baseFile, history);
+}
+
+bool Importer::ImporterImpl::resolveImport(const ImportedEntityPtr &importedEntity,
+                                           const std::string &destination,
+                                           const std::string &typeString,
+                                           const std::string &baseFile,
+                                           std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    if (importedEntity->isImport()) {
+        ImportSourcePtr importSource = importedEntity->importSource();
+        auto h = std::make_tuple(destination, importedEntity->importReference(), importSource->url());
+
+        // Check for cyclical dependencies using the entire tuple contents.
+        if (std::find(history.begin(), history.end(), h) != history.end()) {
+            history.emplace_back(h);
+            return false;
+        }
+        history.emplace_back(h);
+
+        // Only parse import if it doesn't already exist in our importer library.
+        if (!importSource->hasModel()) {
+            std::string url = importSource->url();
+            if (mLibrary.count(url) == 0) {
+                url = resolvePath(importSource->url(), baseFile);
+            }
+            if (mLibrary.count(url) == 0) {
+                // If the url has not been resolved into a model in this library, parse it and save.
+                std::ifstream file(url);
+                if (file.good()) {
+                    std::stringstream buffer;
+                    buffer << file.rdbuf();
+                    auto parser = Parser::create();
+                    auto model = parser->parseModel(buffer.str());
+                    importSource->setModel(model);
+
+                    // Save the pair of model and url to the library map.
+                    mLibrary.insert(std::make_pair(url, model));
+
+                    // Save the imported model to a list of external dependencies.
+                    mExternals.emplace_back(std::make_pair(url, importedEntity->importReference()));
+                    return doResolveImports(model, url, history);
+                }
```

Comment:
Are we really ok with returning `true` if the file cannot be opened? This seems wrong to me.

Reply:
728e78d1